src/include/storage/metric/transaction_metric.h

```diff
@@ -0,0 +1,236 @@
+   */
+  int64_t GetTupleDelete(transaction::timestamp_t txn_id) { return data_[txn_id].tuple_delete_; }
+
+  /**
+   * Get the SQL table for persisting collected data, create a new table if necessary
+   * @param txn_manager transaction manager of the system
+   * @param catalog catalog of the system
+   * @param txn transaction context used for table lookup/creation
+   */
+  catalog::SqlTableHelper *GetStatsTable(transaction::TransactionManager *txn_manager, catalog::Catalog *catalog,
+                                         transaction::TransactionContext *txn) override;
+
+ private:
+  /**
+   * Collection of data related to a transaction
+   */
+  struct TransactionData {
+    std::chrono::high_resolution_clock::time_point start_;
+    int64_t latency_;
+    int64_t tuple_read_;
+    int64_t tuple_insert_;
+    int64_t tuple_delete_;
+    int64_t tuple_update_;
+  };
+
+  /**
+   * Maps from transaction id to relevant data.
+   */
+  std::unordered_map<transaction::timestamp_t, struct TransactionData> data_;
+};
+
+/**
+ * Interface that owns and manipulates TransactionMetricRawData
+ */
+class TransactionMetric : public AbstractMetric<TransactionMetricRawData> {
+ public:
+  /**
+   * @param txn transaction context of the beginning transaction
+   */
+  void OnTransactionBegin(const transaction::TransactionContext *txn) override { GetRawData()->SetTxnStart(txn); }
+
+  /**
+   * @param txn transaction context of the committing transaction
+   * @param database_oid OID of the database the transaction is running in
+   */
+  void OnTransactionCommit(const transaction::TransactionContext *txn, catalog::db_oid_t database_oid) override {
+    (void)database_oid;
+    GetRawData()->CalculateTxnLatency(txn);
+  }
+
+  /**
+   * @param txn transaction context of the aborting transaction
+   * @param database_oid OID of the database the transaction is running in
+   */
+  void OnTransactionAbort(const transaction::TransactionContext *txn, catalog::db_oid_t database_oid) override {
+    (void)database_oid;
+    GetRawData()->CalculateTxnLatency(txn);
+  }
+
+  /**
+   * @param txn context of the transaction performing read
+   * @param database_oid OID of the database that the tuple read happens
+   * @param namespace_oid OID of the namespace that the tuple read happens
+   * @param table_oid OID of the table that the tuple read happens
+   */
+  void OnTupleRead(const transaction::TransactionContext *txn, catalog::db_oid_t database_oid,
+                   catalog::namespace_oid_t namespace_oid, catalog::table_oid_t table_oid) override {
+    (void)database_oid;
+    (void)namespace_oid;
+    (void)table_oid;
+    GetRawData()->IncrementTupleRead(txn);
+  }
+
+  /**
+   * @param txn context of the transaction performing update
+   * @param database_oid OID of the database that the tuple delete happens
+   * @param namespace_oid OID of the namespace that the tuple delete happens
+   * @param table_oid OID of the table that the tuple delete happens
```

Comment:
This and the above parameter comments should say `update` 
```suggestion
   * @param table_oid OID of the table that the tuple update happens
```

I believe it would be better to just have this:
``` suggestion
   * @param table_oid OID of the table
```
And before  any `@param`'s add a comment saying that this function performs actions on tuple updates

Reply:
Corrected